### PR TITLE
lantiq: add 6.1 testing support

### PR DESCRIPTION
**Pull request by sch-m**
@PolynomialDivision This is the continuation of #13046.

What I've done in addition to @hauke's work:

- refreshed all kernel configs and patches for 5.15 and 6.1 once again
- extended the PCIe and GPTU drivers so that they fetch the IRQs from the DeviceTree and as a result eliminated the HACK patch
- removed the last 4 GSWIP patches from the patchset for now due to the resulting functionality issues. The driver needs to be looked at again independently of the linux update.
- fix/work around the problem with the warning regarding field-spanning write in the DSL MEI driver.
- extended the drivers ltq-atm, ltq-ptm, ltq-adsl-mei and ltq-vmmc to fetch the IRQs from the DeviceTree.
- performed tests with xrx200 (VRX268; no voice).

### CODE REVIEW

**Comment by sch-m** [comment]
@PolynomialDivision What can I do to get this PR accepted?

**Comment by abajk** [comment]
~~~
Tested-by: Aleksander Jan Bajkowski <olek2@wp.pl> # Tested on AVM 5490
~~~

File: target/linux/lantiq/ase/config-5.15 [hunk 1]
```diff
@@ -15,7 +14,7 @@ CONFIG_LANTIQ_ETOP=y
 CONFIG_NLS=y
 CONFIG_SGL_ALLOC=y
 CONFIG_SOC_AMAZON_SE=y
-CONFIG_SOC_TYPE_XWAY=y
+# CONFIG_SOC_XWAY is not set
```
Comment (abajk) [inline target/linux/lantiq/ase/config-5.15:17]: I would leave CONFIG_SOC_TYPE_XWAY enabled on the ase subtarget.

File: target/linux/lantiq/ase/config-6.1 [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+CONFIG_GENERIC_ALLOCATOR=y
+CONFIG_GPIO_GENERIC=y
+CONFIG_GPIO_GENERIC_PLATFORM=y
+CONFIG_HW_RANDOM=y
+# CONFIG_ISDN is not set
+CONFIG_LANTIQ_ETOP=y
+CONFIG_NLS=y
+CONFIG_SGL_ALLOC=y
+CONFIG_SOC_AMAZON_SE=y
+# CONFIG_SOC_XWAY is not set
```
Comment (abajk) [inline target/linux/lantiq/ase/config-6.1:17]: Same as above.

**Review by abajk — CHANGES_REQUESTED** [review]

File: target/linux/lantiq/ase/config-5.15 (hunk 1 above)
Comment (sch-m) [reply target/linux/lantiq/ase/config-5.15:17]: The `CONFIG_SOC_TYPE_XWAY` was moved from the subtarget configs to the target config by @hauke:
https://github.com/openwrt/openwrt/pull/15233/commits/7ce95c34b9528ca0f7ddf38f1ef97131816548fb#diff-4674295b90225b63952a5d5a76b57933a6324e19ce2cc8031bee5cf8e8c54c1aR148

Comment (abajk) [reply target/linux/lantiq/ase/config-5.15:17]: You're right, but this is overridden by disabling this symbol on the subtarget.

Comment (sch-m) [reply target/linux/lantiq/ase/config-5.15:17]: No, you have to look closely. These are 2 different symbols: 
`CONFIG_SOC_TYPE_XWAY` and
`CONFIG_SOC_XWAY`.

Comment (sch-m) [reply target/linux/lantiq/ase/config-5.15:17]: @hauke also noted in his commit message that after this refresh the resulting .configs in the kernel tree are still unchanged.

**Comment by PolynomialDivision** [comment]
Did you send some of the kernel patches upstream?

**Comment by abajk** [comment]
WiFi does not work on the AVM 7320 with the new kernel.
6.1 -> https://pastebin.com/nAtrbSgL
5.15 -> https://pastebin.com/QL7GW7bs
USB and Ethernet work the same as on 5.15. Unfortunately, I have no way to test the ADSL2+ part.

In the next step, I plan to do tests on the BT Home Hub5A and see if WiFi on the PCI bus works there.

**Comment by nouman8** [comment]
wifi and vdsl2 work on bt-hh5a i have tested.

**Comment by sch-m** [comment]
> > WiFi does not work on the AVM 7320 with the new kernel.
> 
> Is the PCI bus working with linux 6.1? What's the output of `lspci`? Is the ath9k driver loaded?

Ah, I overlooked the output of lspci at the top of your log. Well, then there seems to be a problem in the PCI bus driver.

**Comment by sch-m** [comment]
Could you please show me the output of `cat /proc/interrupts` with 5.15 and 6.1 running on the AVM 7320?

**Comment by sch-m** [comment]
> wifi and vdsl2 work on bt-hh5a i have tested.

Also interesting that the wifi works on the bt-hh5a. According to dts, there is also an ath9k connected via PCI.

**Comment by abajk** [comment]
> Could you please show me the output of `cat /proc/interrupts` with 5.15 and 6.1 running on the AVM 7320?

A freshly installed image without keeping the configuration:
~~~
root@OpenWrt:/# uname -a
Linux OpenWrt 5.15.155 #0 SMP Wed May 1 09:17:24 2024 mips GNU/Linux
root@OpenWrt:/# cat /proc/interrupts 
           CPU0       CPU1       
  7:    8450694    8421272      MIPS   7  timer
  8:     107991     403379      MIPS   0  IPI call
  9:       1812       1604      MIPS   1  IPI resched
 30:          0          0       icu  30  ath9k
 62:          0          0       icu  62  1e101000.usb, dwc2_hsotg:usb1
 63:     256664          0       icu  63  DFEIR
 72:       2208          0       icu  72  eth_rx
 73:       3809          0       icu  73  eth_tx
 91:          0          0       icu  91  1e106000.usb, dwc2_hsotg:usb2
112:       1563          0       icu 112  asc_tx
113:        121          0       icu 113  asc_rx
114:          0          0       icu 114  asc_err
126:          0          0       icu 126  gptu
127:          0          0       icu 127  gptu
128:          0          0       icu 128  gptu
129:          0          0       icu 129  gptu
130:          0          0       icu 130  gptu
131:          0          0       icu 131  gptu
ERR:          1
~~~

~~~
root@OpenWrt:/# uname -a
Linux OpenWrt 6.1.89 #0 SMP Wed May  1 09:17:24 2024 mips GNU/Linux
root@OpenWrt:/# cat /proc/interrupts 
           CPU0       CPU1       
  7:      13149      12759      MIPS   7  timer
  8:       1919       7495      MIPS   0  IPI call
  9:       1012        897      MIPS   1  IPI resched
 62:          0          0       icu  62  1e101000.usb, dwc2_hsotg:usb1
 63:        142          0       icu  63  DFEIR
 72:        514          0       icu  72  eth_rx
 73:        853          0       icu  73  eth_tx
 91:          0          0       icu  91  1e106000.usb, dwc2_hsotg:usb2
112:        357          0       icu 112  asc_tx
113:         45          0       icu 113  asc_rx
114:          0          0       icu 114  asc_err
126:          0          0       icu 126  gptu
127:          0          0       icu 127  gptu
128:          0          0       icu 128  gptu
129:          0          0       icu 129  gptu
130:          0          0       icu 130  gptu
131:          0          0       icu 131  gptu
ERR:          1
~~~

**Comment by PolynomialDivision** [comment]
@sch-m please send the fixes upstream and backport the patch send upstream. It would be nice even for the patches you keep downstream to write a complete patch header. That makes it easier to directly know what a patch is about in the future. We have also now a kernel bumper script. You could also use that to preserve history. Otherwise I think the changes are fine.  Thanks a lot for your work.

**Comment by sch-m** [comment]
@abajk Maybe we have a problem with this change in 6.1:

https://github.com/torvalds/linux/commit/90c2d2eb7ab5848c4f853751f12b96bdc460ad1b

This reversed the active-low/high logic of the reset GPIO, didn't it?

**Comment by abajk** [comment]
> @abajk Maybe we have a problem with this change in 6.1:
> 
> [torvalds/linux@90c2d2e](https://github.com/torvalds/linux/commit/90c2d2eb7ab5848c4f853751f12b96bdc460ad1b)
> 
> This reversed the active-low/high logic of the reset GPIO, didn't it?

Reverting this change fixes the wifi on the router. I wonder why pci works on some devices and how we can identify them.

**Comment by sch-m** [comment]
According to the specification, PCI RST# is an active low signal. However, in all the related DTS the reset-gpio is defined as active high and the driver has reversed the logic internally up to linux 6.0.

So in my opinion it would actually be correct to change the gpio to active low in the DTS files.

**Comment by sch-m** [comment]
I've added a commit to invert the pci gpio-reset polarity in the dts files.

@abajk @nouman8
Can you please test again on your boards if the wifi works.

**Comment by abajk** [comment]
> @abajk @nouman8 Can you please test again on your boards if the wifi works.

Unfortunately, it seems that wifi still does not work ([logs](https://pastebin.com/4VER2qsz)).

**Comment by abajk** [comment]
During the upgrade, a trace log also appears:
~~~
root@OpenWrt:/tmp# sysupgrade -n openwrt-lantiq-xway-avm_fritz7320-squashfs-sysu
pgrade.bin 
Fri May  3 07:21:27 UTC 2024 upgrade: Commencing upgrade. Closing all shell sessions.
[  653.152811] do_page_fault(): sending SIGSEGV to dnsmasq for invalid read access from 77d6906d
[  653.160158] epc = 77e1f100 in ld-musl-mips-sf.so.1[77ded000+b2000]
[  653.166018] ra  = 77e1f6a4 in ld-musl-mips-sf.so.1[77ded000+b2000]
Watchdog handover: fd=3
- watchdog -
Watchdog did not previously reset the system
Fri May  3 07:21:30 UTC 2024 upgrade: Sending TERM to remaining processes ...
Fri May  3 07:21:30 UTC 2024 upgrade: Sending signal TERM to dsl_cpe_control (2514)
Fri May  3 07:21:34 UTC 2024 upgrade: Sending KILL to remaining processes ...
Fri May  3 07:21:34 UTC 2024 upgrade: Sending signal KILL to dsl_cpe_control (2514)
[  667.034201] stage2 (2730): drop_caches: 3
Fri May  3 07:21:43 UTC 2024 upgrade: Switching to ramdisk...
mount: mounting /dev/mtdblock4 on /overlay failed: Resource busy
[  673.615592] ------------[ cut here ]------------
[  673.618891] WARNING: CPU: 1 PID: 2730 at fs/super.c:503 generic_shutdown_super+0x158/0x198
[  673.627056] VFS: Busy inodes after unmount of jffs2 (jffs2)
[  673.627093] Modules linked in: ath9k ath9k_common nft_fib_inet nf_flow_table_inet ath9k_hw ath pppoe nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_                                                                                                                                                                                                            c
[  673.706066] CPU: 1 PID: 2730 Comm: busybox Not tainted 6.1.89 #0
[  673.711925] Stack : 00000000 80085a14 00000000 00000004 00000000 00000000 83e9bd1c 80a40000
[  673.720206]         80910000 807a8334 80f657d0 80910873 00000000 00000001 83e9bcc8 80c3c800
[  673.728484]         00000000 00000000 807a8334 83e9bbf0 ffffefff 00000000 00000000 ffffffea
[  673.736757]         000000c7 83e9bbfc 000000c7 80890638 807a8334 00000001 83e9bdc8 801d6660
[  673.745049]         00000009 807b0f58 80888258 7f71d870 00000018 00000030 00000004 80a40004
[  673.753328]         ...
[  673.755736] Call Trace:
[  673.758162] [<8000c5ec>] show_stack+0x28/0xf0
[  673.762470] [<806bf578>] dump_stack_lvl+0x60/0x80
[  673.767137] [<80030e10>] __warn+0xb0/0xe4
[  673.771092] [<80030fc4>] warn_slowpath_fmt+0x180/0x188
[  673.776182] [<801d6660>] generic_shutdown_super+0x158/0x198
[  673.781721] [<80432c34>] kill_mtd_super+0x14/0x30
[  673.786361] [<802759ec>] jffs2_kill_sb+0x5c/0x74
[  673.790946] [<801d60f8>] deactivate_locked_super+0x4c/0xbc
[  673.796367] [<801fe638>] cleanup_mnt+0xb0/0x15c
[  673.800870] [<80050e30>] task_work_run+0x94/0xd8
[  673.805425] [<8000bb3c>] do_notify_resume+0x214/0x248
[  673.810446] [<80007690>] work_notifysig+0x10/0x18
[  673.815088] 
[  673.816676] ---[ end trace 0000000000000000 ]---
Fri May  3 07:21:50 UTC 2024 upgrade: Performing system upgrade...
[  673.999026] do_stage2 (2730): drop_caches: 3
Unlocking firmware ...

Writing from <stdin> to firmware ...     
Fri May  3 07:22:49 UTC 2024 upgrade: Upgrade completed
Fri May  3 07:22:50 UTC 2024 upgrade: Rebooting system...
umount: can't unmount /dev: Resource busy
[  734.781170] reboot: Re�
ROM VER: 1.1.3
~~~

It was not present on the 5.15 kernel.

**Comment by nouman8** [comment]
tested with the new "dts Invert" commit on bt-hh5a both ath9k and ath10k wireless are working

**Comment by sch-m** [comment]
> Unfortunately, it seems that wifi still does not work ([logs](https://pastebin.com/4VER2qsz)).

Well, I think the remaining problem is that the new code in the lantiq-pci driver initializes the GPIO in LOW state (independently  of the ACTIVE-LOW/HIGH flag in the dts), whereas the former driver code has set it to HIGH state.

So maybe the best would be to keep the dts files unchanged and patch the driver once again to restore the old behavior but use the newer gpiod API.

**Comment by sch-m** [comment]
@abajk Could you please test the latest version of my patch set?

File: target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9.dtsi [hunk 3]
```diff
@@ -528,7 +528,7 @@
 			reg = <0xd900000 0x1000>;
 
 			interrupt-parent = <&icu0>;
-			interrupts = <161 144>;
+			interrupts = <161 144 145 146 147>;
```
Comment (hauke) [inline target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9.dtsi:531]: Please move this commit before the commit which make use of it.

File: target/linux/lantiq/patches-5.15/0008-MIPS-lantiq-backport-old-timer-code.patch [hunk 4]
```diff
@@ -1005,15 +1007,24 @@ Signed-off-by: John Crispin <blogic@openwrt.org>
 +	}
 +
 +	for (i = 0; i < timer_dev.number_of_timers; i++) {
-+		ret = request_irq(TIMER_INTERRUPT + i, timer_irq_handler, IRQF_TIMER, gptu_miscdev.name, &timer_dev.timer[i]);
```
Comment (hauke) [inline target/linux/lantiq/patches-5.15/0008-MIPS-lantiq-backport-old-timer-code.patch:1008]: If the `TIMER_INTERRUPT` define is not used any more, you can remove it too.

File: target/linux/lantiq/patches-6.1/0002-MIPS-pci-lantiq-restore-reset-gpio-polarity.patch [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+From: Martin Schiller <ms@dev.tdt.de>
+Date: Mon, 6 May 2024 09:41:42 +0200
+Subject: [PATCH] MIPS: pci: lantiq: restore reset gpio polarity
+
+Commit 90c2d2eb7ab5 ("MIPS: pci: lantiq: switch to using gpiod API") not
+only switched to the gpiod API, but also inverted / changed the polarity
+of the GPIO.
+
+According to the PCI specification, the RST# pin is an active-low
+signal. However, most of the device trees that have been widely used for
```
Comment (hauke) [inline target/linux/lantiq/patches-6.1/0002-MIPS-pci-lantiq-restore-reset-gpio-polarity.patch:11]: Lantiq often used the PCI reset pin as active low to support a dying gasp on DSL PCIe cards for example. If the board looses power the reset pin would go to low and the PCIe card can send a dying gasp.

**Comment by hauke** [comment]
Thank you for taking care of these patches. 
Did someone test this on none xrx200 devices?

**Comment by abajk** [comment]
> @abajk Could you please test the latest version of my patch set?

I will test patches this on several devices tomorrow. I have not had access to a computer in the last week.

**Comment by abajk** [comment]
> @abajk Could you please test the latest version of my patch set?

Latest set of patches fixes wifi.

File: target/linux/lantiq/patches-5.15/0008-MIPS-lantiq-backport-old-timer-code.patch (hunk 4 above)
Comment (sch-m) [reply target/linux/lantiq/patches-5.15/0008-MIPS-lantiq-backport-old-timer-code.patch:1008]: The `TIMER_INTERRUPT` is still used in the `timer_irq_handler` to get calculate the timer index out of the irq number at the moment.

**Comment by sch-m** [comment]
> > @abajk Could you please test the latest version of my patch set?
> 
> Latest set of patches fixes wifi.

Thanks for your feedback.

**Comment by sch-m** [comment]
> @sch-m please send the fixes upstream and backport the patch send upstream. It would be nice even for the patches you keep downstream to write a complete patch header. That makes it easier to directly know what a patch is about in the future. We have also now a kernel bumper script. You could also use that to preserve history. Otherwise I think the changes are fine. Thanks a lot for your work.

@PolynomialDivision I have already prepared the patches and will send them upstream as soon as possible. When they are accepted upstream I will replace the patches with the upstream versions.
In addition, all patches should now have a complete header. Furthermore I have now used the kernel bumper script to "copy" the files for 6.1.

**Comment by abajk** [comment]
> > > @abajk Could you please test the latest version of my patch set?
> > 
> > 
> > Latest set of patches fixes wifi.
> 
> Thanks for your feedback.

I think I uploaded the wrong version of the firmware to the device yesterday and my test results were too optimistic :/
I compiled the latest image again and WiFi does not work.
Logs: https://pastebin.com/kt6eSfhJ

**Comment by PolynomialDivision** [comment]
for me it looks good. furthermore, this is just an intermediate step to switch to 6.6 kernel. i think also @hauke comments are resolved. so we can proceed and merge this.

**Comment by sch-m** [comment]
> I think I uploaded the wrong version of the firmware to the device yesterday and my test results were too optimistic :/ I compiled the latest image again and WiFi does not work. Logs: https://pastebin.com/kt6eSfhJ Do you have any new ideas?

@abajk And if you revert the upstream commit instead of using my fix then it works?

I don't see any difference in logic. Can you verify this again?

The only thing that seems a bit inappropriate to me is the `mdelay(1)` for the RESET. This could be a bit tight for some PCI devices. Perhaps you could increase it as a test.

**Comment by PolynomialDivision** [comment]
Can you push a rebase on latest main?

**Review by hauke — APPROVED** [review]

**Comment by PolynomialDivision** [comment]
@sch-m Thanks a lot! Meged. Can you maybe proceed with 6.6 support? :)

**Comment by sch-m** [comment]
> @sch-m Thanks a lot! Merged. Can you maybe proceed with 6.6 support? :)

I'm already on it.

**Comment by Neustradamus** [comment]
@sch-m: Good job!

**Comment by dhewg** [comment]
> > Are some of these changes applicable to VRX518 DSL driver? I can see the field-spanning write in drv_mei_cpe_msg_process.c:3570 in FRITZ!BOX 7530.
> 
> You can port this change to the ltq-vdsl-vr11-mei driver or maybe find a cleaner solution: [392b84a](https://github.com/openwrt/openwrt/commit/392b84aa7f6767c40a2b79654bb205e616b868d3)

Any update on this, or was this patch lost? I'd be nice to fix this warning for both vr9 and vr11 at the same time

**Comment by sch-m** [comment]
> Any update on this, or was this patch lost? I'd be nice to fix this warning for both vr9 and vr11 at the same time

Since this patch set was only lantiq related, I only patched the ltq-vdsl-vr9-mei package. But I can also create a PR for ltq-vdsl-vr11-mei.

**Comment by abajk** [comment]
@sch-m @hauke Do you remember which upstream commit broke the interrupt allocation?

The same issue affects the upstream etop driver. I wrote a patch that fixes it and I am thinking about which kernel versions I need to mark as broken. Linux network mailing list works fast so I expect it will be quickly backported.

**Pull request merged**